<!--
University of Freiburg WS 2017/2018
Chair for Bioinformatics
Supervisor: Martin Raden
Author: Alexander Mattheis
-->

<div id="algorithm_description">
    <div class="description">
        The progressive alignment approach by
        <a href="https://doi.org/10.1007/BF02603120">Da-Fei Feng, Russell F. Doolittle (1987)</a>
        computes a multi-sequence-alignment (MSA) of a set of sequences.
        This algorithm shows an approximative alignment approach,
        which finds considerable solutions in a reasonable time.
        <br>
        <br>
        The below implemented version of this algorithm is a
        combination of several Feng-Doolittle approaches and the affine alignment approach
        by <a href="https://doi.org/10.1016/0022-2836(82)90398-9">Osamu Gotoh (1982)</a>.
        <br>
        After a pairwise alignment of all input sequences
        with <a href="http://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Gotoh">Gotoh</a>
        to compute the similarity-scores,
        this altered <a href="10.1016/s0076-6879(96)66023-6">Feng-Doolittle (1996)</a> approach
        uses a modified formula to transform similarities into distances

        $$\quad d_{a,b} = -ln(s^{eff}_{a,b} \cdot f)$$

        where $f = 1$ (originally $f = 100$)
        <br>
        and
        $$\quad s^{eff}_{a,b} = \frac{s_{a,b} - s^{rand}_{a,b}}{s^{max}_{a,b} - s^{rand}_{a,b}}$$

        where <br>
        <div class="row">
            <div class="colW150"> $s^{max}_{a,b} = \cfrac{s_{a,a}+s_{b,b}}{2}$</div>
            <div class="colW400">average score of aligning both sequences with themselves</div>
	    </div>

        The computation of $s^{rand}_{a,b}$ is done with an approximate formula
        which has been adapted, <br>
        because of the fusion with <a href="https://doi.org/10.1016/0022-2836(82)90398-9">Gotoh</a>.
        The new formula is therefore
        $$
        \quad
        \frac{1}{L_{a,b}}
        \bigg(
        \sum_{i \in \mathbb{A}(a)} \sum_{j \in \mathbb{A}(b)}
        s \Big(i,k | i = k \wedge k \in \mathbb{A}(b) \Big)
        \cdot N_a(i) \cdot N_b(j)
        \bigg)
        + N_{a,b}(\_^e) \cdot \beta
        + N_{a,b}(\_^s) \cdot \alpha
        $$

        where <br>
        <div class="row">
            <div class="colW150">$L_{a,b}$</div>
            <div class="colW400">number of columns in the alignment of $a,b$</div>
	    </div>

        <div class="row">
            <div class="colW150">$\mathbb{A}(a), \mathbb{A}(b)$</div>
            <div class="colW400">alphabet-sets of sequences $a$ and $b$</div>
        </div>

        <div class="row">
            <div class="colW150">$s(i,k | i = k \wedge k \in \mathbb{A}(b))$</div>
            <div class="colW400">similarity between matching characters $i$ and $k$</div>
        </div>

        <div class="row">
            <div class="colW150">$N_a(i), N_b(j)$</div>
            <div class="colW400">frequency of characters $i$ and $j$ in sequences $a$ and $b$</div>
        </div>

        <div class="row">
            <div class="colW150">$N_{a,b}(\_^e)$</div>
            <div class="colW400">frequency of gaps</div>
        </div>

        <div class="row">
            <div class="colW150">$N_{a,b}(\_^s)$</div>
            <div class="colW400">frequency of gap starts</div>
        </div>

        After all similarities have been translated into distances with the given formula,
        a hierarchical clustering algorithm is used to create phylogenetic tree.
        In the original approach, the algorithm
        of <a href="10.1126/science.155.3760.279">Walter M. Fitch and Emanuel Margoliash (1967)</a>
        has been used.
        This implementation instead utilizes the UPGMA algorithm
        by Robert R. Sokal and Charles D. Michener (1958) from the article
        "A statistical method for evaluating systematic relationship".
        The phylogenetic tree is needed as a guide-tree
        that tells the order in which sequences or groups of sequences
        have to be aligned together to create a multi-sequence-alignment (MSA).
        In this connection the "once a gap, always a gap" rule is applied
        by substituting gaps in groups of sequences before aligning them.
        The rule ensures that before made gaps in some subgroup
        of sequences are preserved.
        <br>
        <br>
        For a given input, the stated steps are computed and shown below.
    </div>

    <div class="picture">
        <img src="Feng-Doolittle-120x90.png">
    </div>
</div>