<!--
University of Freiburg WS 2017/2018
Chair for Bioinformatics
Supervisor: Martin Raden
Author: Alexander Mattheis
-->

<div id="algorithm_description">
    <div class="description">
        The dynamic programming approach
        <a href="https://doi.org/10.1016/0022-2836(70)90057-4">Needleman-Wunsch (1970)</a>
        by Saul B. Needleman and Christian D. Wunsch
        computes the optimal, global alignment of two sequences.
        This algorithm from 1970 has used a minimizing scoring-function,
        the maximizing variant was developed 1974 by Peter H. Sellers
        and it is sometimes called
        <a href="https://doi.org/10.1137/0126070">Needleman-Wunsch-Sellers (1974)</a>.
        But both variants can compute the same alignments,
        so they are equivalent with respect to their input and output space.
        <br>
        <br>
        The main motivation for this algorithm is biological and medical examination.
        You want have tools to compare two sequences with each other
        and many algorithms which are used today for alignment are based
        on the
        <a href="https://doi.org/10.1016/0022-2836(70)90057-4">Needleman-Wunsch</a>
        algorithm.
        <br>
        <br>
        In the first step of Needleman-Wunsch(-Sellers) you initialize a matrix $D$.
        Then you fill that matrix with the help of a recursion function $D_{i,j}$.
        And in the last step you do matrix tracebacks to compute the alignments
        of the two input-sequences $S_1$ an $S_2$. The runtime and the space costs are in
        $\mathcal O(nm)$, where $n=|S_1|$ and $m=|S_2|$.
    </div>

    <div class="picture">
        <img src="Needleman-Wunsch-120x90.png">
    </div>
</div>

<div id="algorithm_interface"></div>