<!--
University of Freiburg WS 2017/2018
Chair for Bioinformatics
Supervisor: Martin Raden
Author: Alexander Mattheis
-->

<div id="algorithm_description">
    <div class="description">
        The dynamic programming approach
        <a href="https://doi.org/10.1016/0022-2836(81)90087-5">Smith-Waterman (1981)</a>
        by Temple F. Smith and Michael S. Waterman
        computes unlike Needleman-Wunsch not
        the optimal, global alignment of two sequences, but the optimal, local alignment.
        This means that no more the whole sequences are aligned,
        but parts of both sequences such that the score is going to be optimal.
        Like in
        <a href="https://doi.org/10.1016/0022-2836(70)90057-4">Needleman-Wunsch (1970)</a>
        and other alignment algorithms,
        there exists two variants. One with minimizing scoring-function
        and also an equivalent maximizing variant.
        <br>
        <br>
        The motivation is the same like in
        <a href="https://doi.org/10.1016/0022-2836(70)90057-4">Needleman-Wunsch</a>,
        you want do biological and medical examination.
        And therefore you need tools to compare
        two sequences with each other.
        Many present approaches are based on the
        <a href="https://doi.org/10.1016/0022-2836(81)90087-5">Smith-Waterman</a> algorithm.
        <br>
        <br>
        In the first step of
        <a href="https://doi.org/10.1016/0022-2836(81)90087-5">Smith-Waterman</a>
        you initialize the left and the upper borders of a matrix with zeros.
        Then, like in
        <a href="https://doi.org/10.1016/0022-2836(70)90057-4">Needleman-Wunsch</a>
        you fill the matrix
        with the help of a recursion function $D_{i,j}$.
        The only difference to the
        <a href="https://doi.org/10.1016/0022-2836(70)90057-4">Needleman-Wunsch</a>
        recursion function
        is that you have a fourth case with a zero in your function.
        And the last step is the traceback-step in which you compute the alignments
        of the two entered sequences $S_1$ and $S_2$.
        Runtime and space
        are again like in the
        <a href="https://doi.org/10.1016/0022-2836(70)90057-4">Needleman-Wunsch</a>
        algorithm,
        so in $\mathcal O(nm)$, where $n=|S_1|$ and $m=|S_2|$.
    </div>

    <div class="picture">
        <img src="Smith-Waterman-120x90.png">
    </div>
</div>

<div id="algorithm_interface"></div>